juju/juju
# S3 Controller Client Worker

Adds a new objectstore s3 caller worker. Originally, the intention was to reuse the s3caller and add the objectstore.Session to that directly. The s3caller worker is run on all unit agents, and we want an objectstore version on the controller. The complexity of the cross-cutting concerns made this untenable. Instead, I've created a new objectstore s3caller. The new worker understands all the requirements from just the object store directly.

The new interface for the object store s3 caller takes cues from the dbaccessor. In that, it expects to be called with a function.

```go
var reader io.Reader
err := client.Session(ctx, func(ctx context.Context, s objectstore.Session) error {
    var err error
    reader, err = s.GetObject(ctx, "foo", "bar")
    return err
})
// do something with the err and reader.
```

With this setup, if the underlying session credentials become invalid, we now have to the ability to retry directly for the consumer. The intention is to try and keep other dependent workers up and only fail once we've tried at least a few times. All requests are not serialized, and we only retry to create a new client if the controller config changes are witnessed. We could move to serializing all writes to the s3client, but we would probably need a client per namespace (model + controller). This feels a bit heavy-handed atm.

This PR also adds a lot of clean up + refactoring of the origin s3caller.

## Checklist




## QA steps

Although we're not using the new client directly, we need to ensure that we're not preventing the dependency engine from starting.

```sh
$ juju bootstrap lxd test --build-agent
```


## Links


**Jira card:** JUJU-5307
